### PR TITLE
Read indices from a shard 

### LINKED ISSUE #2459: Fail to fetch shard using multi-process in Python.
```python
class ImageDataset(IterableDataset):
    def __init__(self, data_shard_service, shuffle=False):
        self.data_shard_service = data_shard_service
        self._shuffle = shuffle

    def __iter__(self):
        while True:
            index = self.data_shard_service.fetch_shard()
            image, label = read_images(index)
            yield image, label

dataset = ImageDataset(
      allreduce_controller.data_shard_service, shuffle=True
  )

data_loader = DataLoader(
      dataset=dataset,
      batch_size=args.batch_size,
      num_workers=2
  )
```

image

### PR DESCRIPTION
**Pull request by workingloong**
Fix #2459

### CODE REVIEW

File: elasticai_api/common/data_shard_service.py [hunk 1]
```diff
@@ -125,12 +130,45 @@ def report_batch_done(self, batch_size=None, err_msg=""):
             return True
         return False
 
-    def fetch_shard(self):
```
Comment (brightcoder01) [inline elasticai_api/common/data_shard_service.py]: TF 1.x custom training loop relies on this method. We can keep this method and also provide the implementation of multiple process.

Comment (workingloong) [reply elasticai_api/common/data_shard_service.py]: Yes, it is a mistake and I have restored it.

File: elasticai_api/common/data_shard_service.py [hunk 2]
```diff
@@ -134,3 +137,51 @@ def fetch_shard(self):
...
+        shuffle=False,
+    ):
+        super(RecordIndexService, self).__init__(
+            master_client, batch_size, num_epochs, dataset_size, task_type
+        )
+        self._shuffle = shuffle
+        self._shard_queue = SimpleQueue()
+        threading.Thread(
+            target=self._get_shard_indices,
+            name="check_timeout_tasks",
```
Comment (skydoorkai) [inline elasticai_api/common/data_shard_service.py]: name not related?

File: elasticai_api/common/data_shard_service.py [hunk 3]
```diff
@@ -134,3 +137,51 @@ def fetch_shard(self):
...
+            daemon=True,
+        ).start()
+
+    def _get_shard_indices(self):
+        while True:
+            if self._shard_queue.empty():
```
Comment (skydoorkai) [inline elasticai_api/common/data_shard_service.py:165]: Don't need to wait for empty, if the queue size is small, such as   `< batchsize`, get next shard.

```
 if self._shard_queue.qsize() < batchsize:

**Review by skydoorkai — APPROVED** [review]

**Pull request merged**